src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js

```diff
@@ -1,171 +1,560 @@
-            var xAxis = svg.append("g")
-                .attr("transform", "translate(0," + height + ")")
-                .call(d3.axisBottom(x));
+        let brush = d3.brush()
+            .extent([[0,0],[box_width,box_height]])
+            .on("start " + brushClass,updateChart)
 
-            var y = d3.scaleLinear()
-                .domain([0, 9])
-                .range([height, 0]);
-            svg.append("g")
-                .call(d3.axisLeft(y));
+        // The brush element must be child of zoom element - this is only way we found to have both zoom and brush
+        // events working at the same time. See https://stackoverflow.com/a/59757276 .
+        let brushElem = zoom.zoomElem.append("g")
+            .attr("class",brushClass)
+            .call(brush)
 
-            var clip = svg.append("defs").append("svg:clipPath")
-                .attr("id", "clip")
-                .append("svg:rect")
-                .attr("width", width)
-                .attr("height", height)
-                .attr("x", 0)
-                .attr("y", 0);
+        let self = this;
 
-            var color = d3.scaleOrdinal()
-                .domain(colorDomain)
-                .range(colorRange)
+        /**
+         * Zooms into selected fragment of plot.
+         */
+        const zoomIn = () => {
+            let xMin = scaleAndAxis.xScale.invert(extent[0][0]);
+            let xMax = scaleAndAxis.xScale.invert(extent[1][0]);
+            let yMin = scaleAndAxis.yScale.invert(extent[1][1]);
+            let yMax = scaleAndAxis.yScale.invert(extent[0][1]);
 
-            var brush = d3.brushX()
-                .extent([[0, 0], [width, height]])
-                .on("end", updateChart)
+            scaleAndAxis.xScale.domain([xMin,xMax]);
+            scaleAndAxis.yScale.domain([yMin,yMax]);
 
-            var scatter = svg.append('g')
-                .attr("clip-path", "url(#clip)")
+            self.zoomingHelper(scaleAndAxis,box_width,scatter,points);
 
-            scatter
-                .selectAll("circle")
-                .data(_data)
-                .enter()
-                .append("circle")
-                .attr("cx", function (d) {
-                    return x(d[headerNames[0]]);
-                })
-                .attr("cy", function (d) {
-                    return y(d[headerNames[1]]);
-                })
-                .attr("r", 8)
-                .style("fill", function (d) {
-                    return color(d[headerNames[2]])
-                })
-                .style("opacity", 0.5)
-
-            scatter
-                .append("g")
-                .attr("class", "brush")
-                .call(brush);
-
-            var idleTimeout
-
-            function idled() {
-                idleTimeout = null;
+            brushElem.call(brush.move,null);
+        }
+
+        const zoomInKeyEvent = (event) => {
+            if ((event.ctrlKey || event.metaKey) && event.key === 'z') {
```

Comment:
refactor all these shortcuts to a shortcut section on top of file (below sconstants section). Like:

```js
let shortcuts = {
    zoomIn : (e) => ((e.ctrlKey || e.metaKey) && e.key === 'z'),
    ...
}
```